FortranCon 2021 stdlib presentation (#21)
fortran-lang/talks: FortranCon2021-stdlib/stdlib-talk.tex

\documentclass[aspectratio=169]{beamer}

\usepackage{ulem}
\usepackage{dejavu}
\usepackage{listings}

\mode<presentation>{
  \usetheme{Rochester}
}
\beamertemplatenavigationsymbolsempty
\usecolortheme[RGB={115,79,150}]{structure}
%\usefonttheme{structurebold}

\title{What's new in the Fortran standard library?}
\author{Nathaniel Shaffer \\[1em] \small
  \and Gabriel Brown
  \and Ond\v{r}ej \v{C}ertik
  \and William Clodius \\
  \and Milan Curcic
  \and Laurence Kedward
  \and Sebastian Ehlert \\
  \and Gareth Davies
  \and Aman Godara
  \and Michael Hirsch \\
  \and Jing
  \and Chetan Karwa 
  \and Arjen Markus
  \and Ivan Pribec \\
  \and Harris Snyder
  \and J\'er\'emie Vandenplas
  \and Evan Voyles}
\date{FortranCon 2021 - 24 Sept 2021}

\begin{document}

\begin{frame}

>   \titlepage

Reviewer: I'm not familiar with Beamer. Would it be easy to add our purple F logo to the title slide? Perhaps to the left of the author list?

Reviewer 2: I added the logo to the `\institute{}` which displays it between the author list and the date.

https://github.com/fortran-lang/talks/blob/d475ff20eb30a965790a1ae89d02fbe78f10ba81/FortranCon2021-fpm/main.tex#L40

Author: Thanks for the tip!

\end{frame}

\begin{frame}

>   \frametitle{What is \sout{new in} stdlib?}

Reviewer: Do you mean this to be a "What is Fortran standard library" slide? In other words, introducing stdlib to people who never heard about it?

If yes, then I suggest adding a bullet a-la:

* Make Fortran easier to use and more powerful (almost) out of the box

or similar.

The "bridge the gap" bullet point kinda communicates that message, but it may not be obvious what it means, at least IMO.

Reviewer: We should also say that one of the goals of stdlib is to prototype future standard intrinsics, and provide early implementations before they're available in compilers.

  \begin{itemize}
  \item Part of fortran-lang: https://github.com/fortran-lang/stdlib
  \item Bridge the gap between intrinsics and applications
    \begin{description}[Mathematics]
    \item[Utilities] strings, logging, filesystem interaction
    \item[Algorithms] searching, sorting
    \item[Mathematics] linear algebra, special functions, statistics
    \end{description}

>   \item Somewhere between C++ stdlib \& Matlab/scipy

Reviewer: I think this is a "scope of stdlib" bullet point, so I suggest specifying that, and also naming the scopes:

```suggestion
  \item Scope: General purpose (e.g. C++, Python) and scientific and engineering (e.g. MATLAB, SciPy) applications
```

But it may be quite busy now. Maybe there's a happy medium

  \end{itemize}

  \note{
    Before I can show off what's new in the standard library, let's have a few words about what it is.
    It's a volunteer effort maintained under the fortran-lang organization, with the code hosted on Github.
    The main goal is to bridge the gap between what's offered intrinsically in the language versus the high-level code one would like to write in applications.
    This makes the scope of the Fortran standard library quite wide.
    We want to offer general programming-type things like common algorithms a la C++, while also providing a good base of numerically-oriented functionality like what's found in Matlab or the Python scientific ecosystem.
    That's a tall order, but we've made good headway.
  }
\end{frame}

> \begin{frame}

Reviewer: I love this slide. What do you think about adding a bullet point under the table that summarizes current count of things, for example:

* ?? derived types and ?? new procedures across 18 modules

  \frametitle{stdlib has roughly doubled in size in the past year}

>   \begin{block}{Modules one year ago vs today}

Reviewer: Red color may mean something bad or negative. I suggest coloring new modules in green.

Reviewer 2: I usually use `\alert{}` for highlighting the color will be orange to yellowish and stand out quite nicely.

      quadrature &
      \textcolor{red}{sorting} &
      \textcolor{red}{specialfunctions} \\            
      stats &
      \textcolor{red}{stats\_distribution\_PRNG} &
      \textcolor{red}{stringlist\_type} \\
      \textcolor{red}{strings} &
      \textcolor{red}{string\_type} &                           
      system 
    \end{tabular}
  \end{block}

  \note{
    By raw count of modules included in stdlib, the project has roughly doubled in size in the past year.
    A year ago, stdlib looked more like a grab-bag of utilities than a standard library.
    Now, it looks much more deserving of its name, with lots of exciting development in general-purpose programming functionality like string handling, sorting, logging, and support for bitsets.
    Let's have a look at some examples.
  }
\end{frame}

\begin{frame}[fragile]
  \frametitle{Demo: stdlib\_logger}

  \begin{block}{ex\_logger.f90}
    \small
\begin{verbatim}
use stdlib_logger, only: global_logger
implicit none
call global_logger%add_log_file('log.txt')
call global_logger%log_debug('I am invisible')
call global_logger%log_information('Something informative')
call global_logger%log_error('Oopsie daisy')

> end

Reviewer 2: Maybe make use of `lstlisting` to get some basic syntax hints in (mainly by typeface not color). For the fpm presentation I highlighted the code snippets externally and included them as pdf.

Author: Hm, I tried mimicking the example in the beamer manual and made a mess of it. In your opinion are the code listings hard to read without any styling?

b2 = [(.true., i=1,6)]        ! S6B111111

call xor(b1, b2)              ! S6B110001, S6B111111

call b1%set(2, 4)             ! S6B111111 -- N.B. 0-based index
print *, b1 == b2             ! T
end
\end{verbatim}
  \end{block}

  \note{
    The second demo is of the bitsets module.
    This provides a feature that is often proposed for standardization: an abstract type for working with strings of bits.
    The basic bitset type uses a 64-bit integer internally, but it keeps track of the actual length of the bitset being represented.
    Here, I'm initializing two 6-bit sets, one from a string and one from an array of logicals.
    The comment shows a string representation of the bitset, which can be taken as a bitset 'literal' for the purposes of I/O.
    Here, the prefix 'S6B' means a set of 6 bits.
    Most operations with bitsets work in-place rather than creating new instances.
    For instance, all the usual boolean algebra operations -- like the exclusive-or here -- are implemented as subroutines that overwrite the first argument.
    Likewise, when one sets or clears single bits or contiguous ranges of bits, the data is modified without making a copy.
  }
\end{frame}

> 

Reviewer: What do you think about adding a slide with `string_type` examples? I think that could be exciting for the audience, it's something that's been a weak spot of standard Fortran.

Author: I expect @Aman-Godara will focus on strings in his lightning talk, and I didn't want to step into his space. I could have something very brief, just enough to highlight it and say that details will follow in Aman's talk.

Reviewer: Ah, yes, you're right. All good.

\begin{frame}
  \frametitle{New contributors have been key to stdlib's growth}

  \begin{itemize}

>   \item From 16 committors to 34, including our 2 GSoC students

Reviewer 2: Better use https://fortran-lang.org/community/ to count contributors, since it includes those participating in the discussions at the issue tracker and the code review. The actual code contributors can be a separate point.

Author: Thanks! I was trying to find this on GitHub, forgetting that it's on our own website!

    Not only has the number of modules doubles but so has the number of people who've committed changes to stdlib.
    Included in that are two Google Summer of Code students Aman Godara and Chetan Karwa, who we will hear from shortly.
    What that figure doesn't include is contributions made in the form of Issues and reviews of Pull Requests.
    It's really great to see over 100 new Issues opened in the past year, since this not only includes bug reports but also discussion of workflow improvments and proposals for new features.
    And looking at the timeline of commits since last year's FortranCon, we see a nice sustained development with big rushes of new development at the end of last summer and throughout this summer.
  }
\end{frame}

\begin{frame}
  \frametitle{stdlib is now easier to install}

  \begin{itemize}
  \item Dependencies
    \begin{itemize}
    \item Fortran compiler (supporting at least F2008)
    \item CMake (or just make)
    \item fypp preprocessor (python script)
    \end{itemize}
  \item Install each separately or use conda package manager
  \item Exports both CMake package files \& pkg-config files

>   \item Support for fpm-based workflow is in progress

Reviewer 3: fpm supports now stdlib using a specific branch of the stdlib project (since a couple of days).
See [here](https://github.com/fortran-lang/stdlib#build-with-fortran-langfpm) for moe details.

Author: Yes! I saw this shortly after opening this PR. I will update that slide accordingly. What great timing!

Reviewer: Yes, consider also making it a short separate slide that would have snippet showing how easy it is to use stdlib as an fpm dependency:

```
[dependencies]
stdlib = { git="https://github.com/fortran-lang/stdlib", branch="stdlib-fpm" }
```

    So the has three dependencies: a reasonably modern Fortran compiler, a reasonably recent version of CMake (or just make if you prefer), and the fypp preprocessor, which is used mainly as a templating engine for writing generic procedures.
    These are individually not hard to install on most systems, but one might be worried about having conflicting versions of Fortran compilers or CMake on one system.
    So to address that concern and also to just serve as a very simple way to get started, there's now a conda package that will automatically install the prerequisites for stdlib.
    So then once you get the stdlib source code from Github, you can build stdlib and run its test suite with either make or CMake.
    Then when you isntall stdlib, you get not only a static or shared library but also a pkg-config file and a CMake package file, which makes it very simply to use stdlib in your own projects.
    The ideal situation of course is to distribute stdlib as an fpm package, and while that's still a work in progres, there's been good progress on the fpm side of things.
  }
\end{frame}

\begin{frame}

>   \frametitle{Cross-platform support monitored with Github's CI workflow}

Reviewer: ```suggestion
  \frametitle{Cross-platform support monitored with GitHub's CI workflow}
```

>   \begin{block}{Patforms tested on every pull request}

Reviewer: ```suggestion
  \begin{block}{Platforms tested on every pull request}
```

    Another issue is that many numerical algorithms should be reviewed by people with particular domain knowledge to ensure that the implementation and tests make sense.
    The prime example of this right now is a handful of pull requests for statistical distributions that have been waiting for a reviewer with an appropriate background in statistics.
    When discussing new feature proposals, a recurring question is, ``should this be in stdlib or should this be an fpm package?''
    Really, this is a testament to the progress made with fpm, but the point is that this boundary is still being mapped out.
    The other area where the stdlib project is actively trying to improve things is in the documentation.
    Even though documentation is required for every new procedure or type, the style and rigor of this documentation varies quite a lot.
    The proposed solution right now is to establish a baseline template for specifications to follow.
    This will hopefully give the stdlib documentation a more uniform feel, as well as be a helpful starting point for new contributors.
  }
\end{frame}

> 

Reviewer: Here's a draft roadmap slide, largely based on the current backlog of PRs, which I think is realistic. Please adapt it as you see fit.

```suggestion

\begin{frame}
  \frametitle{Outlook: Next 12 months}

  \begin{itemize}
  \item Probability distributions: Uniform, normal, exponential, gamma, and beta (Fisher)
  \item Generic linked list (Karwa)
  \item Generic map type (Ehlert)
  \item Hash functions (Clodius, Ehlert)
  \item Improved OS and file system facilities (Diehl, Markus)
  \item Selection algorithms
  \item All new intrinsics planned for Fortran 202X
  \item Improved stdlib test suite (Ehlert)
  \end{itemize}

  \note{Here are some additions and improvements to look forward to in the next year or so.
        The highlights are: procedures for various probability distributions and sampling
        pseudo-random numbers from them; derived types for generic collections
        (linked lists, see Chetan's talk today, and a map/dictionary type);
        and cross-platform procedures for querying the operating system and manipulating
        the file system (such as listing and traversing directories, removing or copying files, etc.).
        This roadmap is driven by community and their contributions and is largely based on
        the backlog of existing PRs to be reviewed and improved upon.
        If there's something you'd like to be a part of stdlib, please let us know by opening
        a proposal in GitHub issues, and we'll discuss it and come up with an implementation plan.
  }

\end{frame}

```

Reviewer 3: Thank you @milancurcic for this roadmap. I would suggest to add to this list: "portability across platforms", or something similar.

\begin{frame}
  \frametitle{Summary}

  \begin{itemize}

>   \item stdlib aims to be a de factor standard library of general-purpose and numerical facilities for Fortran

Reviewer 4: ```suggestion
  \item stdlib aims to be a de facto standard library of general-purpose and numerical facilities for Fortran
```

Otherwise, it's got the "okay" from me. Very sharp looking and informative presentation, great job!

Author: Right-o, good catch!

    Likewise, when one sets or clears single bits or contiguous ranges of bits, the data is modified without making a copy.
  }
\end{frame}

\begin{frame}
  \frametitle{New contributors have been key to stdlib's growth}

  \begin{itemize}
  \item From 16 committors to 34, including our 2 GSoC students
  \item Over 100 new Issues: bugs, workflow improvements, feature proposals
  \item From 52 to 97 contributors (commits, discussion, reviews)
  \end{itemize}

  \begin{block}{Commits to stdlib since FortranCon 2020}
    \centering
    \includegraphics[width=\textwidth]{contrib.png}
  \end{block}

  \note{
    So we got lots of great new functionality, and in large part that's thanks to an influx of new contributors to the stdlib project.
    Not only has the number of modules doubles but so has the number of people who've committed changes to stdlib.
    Included in that are two Google Summer of Code students Aman Godara and Chetan Karwa, who we will hear from shortly.
    What that figure doesn't include is contributions made in the form of Issues and reviews of Pull Requests.

>     It's really great to see over 100 new Issues opened in the past year, since this not only includes bug reports but also discussion of workflow improvments and proposals for new features.

Reviewer 5: ```suggestion
    It's really great to see over 100 new Issues opened in the past year, since this not only includes bug reports but also discussion of workflow improvements and proposals for new features.
```

    And looking at the timeline of commits since last year's FortranCon, we see a nice sustained development with big rushes of new development at the end of last summer and throughout this summer.
  }
\end{frame}

\begin{frame}
  \frametitle{stdlib is now easier to install}

  \begin{itemize}
  \item Dependencies
    \begin{itemize}
    \item Fortran compiler (supporting at least F2008)
    \item CMake (or just make)
    \item fypp preprocessor (python script)
    \end{itemize}
  \item Install each separately or use conda package manager
  \item Exports both CMake package files \& pkg-config files
  \item New support for fpm-based workflow
  \end{itemize}

  \note{

>     Besides implementing and discussing new features, the other way in which stdlib has improve a lot this past year is in infrastructure.

Reviewer 5: ```suggestion
    Besides implementing and discussing new features, the other way in which stdlib has improved a lot this past year is in infrastructure.
```

    So to address that concern and also to just serve as a very simple way to get started, there's now a conda package that will automatically install the prerequisites for stdlib.
    So then once you get the stdlib source code from Github, you can build stdlib and run its test suite with either make or CMake.
    Then when you install stdlib, you get not only a static or shared library but also a pkg-config file and a CMake package file, which makes it very simply to use stdlib in your own projects.
    But the ideal is to distrubute stdlib as an fpm package, which is newly supported!
  }
\end{frame}

\begin{frame}[fragile]
  \frametitle{It is now trivial for fpm packages to depend on stdlib}

  \begin{block}{fpm.toml}
    \small
\begin{verbatim}
...
[dependencies]
stdlib.git = "https://github.com/fortran-lang/stdlib"
stdlib.branch = "stdlib-fpm" 
...
\end{verbatim}
  \end{block}

  \begin{center}
    It just works!
  \end{center}

  \note{
    For a long time, stdlib's preprocessing needs were difficult to accomodate with fpm.

>     Now we maintain a branch of stdlib that just contains the pre-processed Fortran source, exacly the way fpm wants it.

Reviewer 5: ```suggestion
    Now we maintain a branch of stdlib that just contains the pre-processed Fortran source, exactly the way fpm wants it.
```

    When discussing new feature proposals, a recurring question is, ``should this be in stdlib or should this be an fpm package?''
    Really, this is a testament to the progress made with fpm, but the point is that this boundary is still being mapped out.
    The other area where the stdlib project is actively trying to improve things is in the documentation.
    Even though documentation is required for every new procedure or type, the style and rigor of this documentation varies quite a lot.
    The proposed solution right now is to establish a baseline template for specifications to follow.
    This will hopefully give the stdlib documentation a more uniform feel, as well as be a helpful starting point for new contributors.
  }
\end{frame}

\begin{frame}
  \frametitle{Outlook: Next 12 months}

  \begin{itemize}
  \item Probability distributions: Uniform, normal, exponential, gamma, and beta (Fisher)
  \item Generic linked list (Karwa)
  \item Generic map type (Ehlert)
  \item Hash functions (Clodius, Ehlert)
  \item Improved OS and file system facilities (Diehl, Markus)

>   \item Selection algorithms

Reviewer: I just saw that the selection algorithms PR is led by @gareth-nx 

```suggestion
  \item Selection algorithms (Davies)
```

Reviewer 3: Do we need to mention names next to the different topics? I have the feeling that it may stop people who will listen to the presentation to participate to one of these topics, because names are already attached to that. However, we need help for stdlib, and even for these topics!

Reviewer 2: I agree with Jeremie on this. Most of these projects are collective efforts already. For example the test suite refactoring is work done by three contributors now and only one is listed here.

Reviewer: I don't think we need it--I thought it'd be nice mentioning the leads, but I agree with both your arguments.

Author: I think we're all on the same page about this. I've struck the "assignees" from this slide.

  \item Portability across platforms
  \item All new intrinsics planned for Fortran 202X
  \item Improved stdlib test suite (Ehlert)
  \end{itemize}

  \note{
    Here are some additions and improvements to look forward to in the next year or so.

>     The highlights are: procedures for various probability distributions andsampling pseudo-random numbers from them;

Reviewer 5: ```suggestion
    The highlights are: procedures for various probability distributions and sampling pseudo-random numbers from them;
```
There seems to be a space missing here.

    derived types for generic collections (linked lists, see Chetan's talk today, and a map/dictionary type);
    and cross-platform procedures for querying the operating system and manipulating the file system
    (such as listing and traversing directories, removing or copying files, etc.).
    This roadmap is driven by community and their contributions and is largely based on the backlog of existing PRs to be reviewed and improved upon.
    If there's something you'd like to be a part of stdlib, please let us know by opening a proposal in GitHub issues, and we'll discuss it and come up with an implementation plan.
  }
\end{frame}

\begin{frame}
  \frametitle{Summary}

  \begin{itemize}
  \item stdlib aims to be a de facto standard library of general-purpose and numerical facilities for Fortran
  \item Roughly doubled in size in the past year, both in terms of modules and contributors
  \item New modules include bitsets, logging, math utilities, sorting, special functions, RNG, and string handling

>   \item Infrastructure and packaging improvments have made stdlib easier to install and use

Reviewer 6: ```suggestion
  \item Infrastructure and packaging improvements have made stdlib easier to install and use
```
Just caught this typo.
